src/test/java/com/spotify/dns/DnsSrvWatchersTest.java

```diff
@@ -0,0 +1,90 @@
+package com.spotify.dns;
+
+import com.google.common.collect.ImmutableList;
+
+import org.junit.Test;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.not;
+
+public class DnsSrvWatchersTest {
+
+  @Test
+  public void triggerRace() throws Exception {
+    int limit = 100000;
+    while (limit-- > 0) {
+      System.out.println(limit);
+      resolve();
+    }
+  }
+
+  private void resolve() throws Exception {
```

Comment:
Actually, I misremembered. The test used to hang. Now it fails with high probability for the earlier version of AbstractChangeNotifier. I couldn't figure out a set to deterministically trigger the test without rewriting in some non-intuitive fashion.